perceval/components/core_catalog/controlled_rotation_gates.py

```diff
@@ -0,0 +1,136 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+import math
+from warnings import warn
+
+from perceval.components import ACircuit, Circuit, Port, Unitary, Processor
+from perceval.components.component_catalog import CatalogItem, AsType
+from perceval.utils import Encoding, PostSelect, Matrix
+
+import numpy as np
+import cmath as cm
+from scipy.linalg import block_diag
+
+
+def build_control_gate_unitary(n: int, alpha: float) -> Matrix:
+    """ Builds the Unitary transformation of the n-qubit controlled-rotation gate C...CZ(alpha)
+
+    :param n: number of qubits of the gate
+    :param alpha: rotation angle of the gate
+    :return: Unitary matrix of the post-selected gate
+             Ref : https://arxiv.org/abs/2405.01395
+       
+    """
+
+    I = Matrix.eye(n)
+    a = (cm.exp(alpha * 1j) - 1) ** (1 / n)
+    J = np.roll(Matrix.eye(n), shift=1, axis=1)
+
+    # Main matrix block
+    A0 = I + a * J
+    B0 = Matrix.eye(n)
+    M = block_diag(B0, A0)
+
+    U0 = Matrix.get_unitary_extension(M)
+
+    # Setting the modes right for the dual rail encoding
+    initial_modes = [i for i in range(4*n)]
+    final_modes = [
+        2*i for i in range(n)] + [2*i+1 for i in range(n)] + [i for i in range(2*n, 4*n)]
+    Perm = Matrix.zeros((4*n, 4*n))
+    Perm[initial_modes, final_modes] = 1
+
+    U = Perm.T @ U0 @ Perm
+
+    return U
+
+
+class PostProcessedControledRotationsItem(CatalogItem):
+    article_ref = "https://arxiv.org/abs/2405.01395"
+    description = r"""n-qubit controlled roation gate C...CZ(alpha) with 2n heralded modes and a post-selection function"""
+    str_repr = r"""                       ╭─────╮
+ctrl0 (dual rail) ─────┤     ├───── ctrl0 (dual rail)
+                  ─────┤     ├─────
+                       │     │
+ctrl1 (dual rail) ─────┤     ├───── ctrl1 (dual rail)
+                  ─────┤     ├─────
+                          .
+                          .
+                          .
+data (dual rail)  ─────┤     ├───── data (dual rail)
+                  ─────┤     ├─────
+                       ╰─────╯"""
+
+    def __init__(self):
+        super().__init__("postprocessed controlled gate")
+        self._default_opts['type'] = AsType.PROCESSOR
+
+    def build_circuit(self, **kwargs):
+        """
+        kwargs:
+            - n : int,  number of qubit of the gate.
+            - alpha : float,  angle of the controlled-rotation.
+
+        :return: Circuit implementing the post-selected n-qubit controlled gate.
+        """
+        if not "n" in kwargs:
+            raise KeyError("Missing input n")
+        n = kwargs["n"]
+        if n<2 :
+            raise ValueError(f"n must be at least 2. Here n = {n}.")
+        if type(n) != int:
```

Comment:
You should test the type before the value of n, if n is a type that cannot get compared with an int, you will have a weird error else.
Also the pythonic way of testing the type is `isinstance(a, type)` can you change you check to this ?